[mono] Implement public hot reload API

lambdageek opened the pull request:
Also add ApplyUpdateSdb that takes byte[] arguments for use with
mono/debugger-libs as its awkward to call a ROS<byte> method from the debugger.

Contributes to https://github.com/dotnet/runtime/issues/45689

This is the mono counterpart of #48366

lambdageek committed:
[mono] Implement public hot reload API

Also add ApplyUpdateSdb that takes byte[] arguments for use with
mono/debugger-libs as its awkward to call a ROS<byte> method from the debugger.

Contributes to https://github.com/dotnet/runtime/issues/45689

lambdageek commented:
Probably going to conflict with the CoreCLR PR since we're both modifying the `ref/` assembly, but that's ok, I'll rebase after the other PR goes in.

lambdageek commented:
@joj  once this is merged, (and there's a new nightly runtime installer) you will need to update your prototype to use
```csharp
System.Reflection.Metadata.AssemblyExtensions.ApplyUpdateSdb(Assembly assembly, byte[] metadataDelta, byte[] ilDelta, byte[] pdbDelta)
```

stephentoub commented on src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs:
```diff
@@ -10,6 +10,7 @@ public static partial class AssemblyExtensions
     {
         [System.CLSCompliantAttribute(false)]
         public unsafe static bool TryGetRawMetadata(this System.Reflection.Assembly assembly, out byte* blob, out int length) { throw null; }
+        public static void ApplyUpdate(System.Reflection.Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta = default) {  }
```
I assume this is just temporary as https://github.com/dotnet/runtime/pull/48366 is adding this as well (but slightly different).

lambdageek replied:
Yep.  Once the other PR goes in, this will be removed.

stephentoub commented on src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:
```diff
@@ -1,11 +1,69 @@
+        /// <param name="pdbDelta">The PDB changes to be applied.</param>
+        /// <exception cref="ArgumentNullException">The assembly argument is null.</exception>
+        /// <exception cref="NotSupportedException">The update could not be applied.</exception>
+        public static void ApplyUpdate(Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta = default)
```
Once https://github.com/dotnet/runtime/pull/48366 is in, can we refactor this to have most of this method in the shared partition?  Seems like the main difference is just the name of the ApplyUpdate_internal helper; we should be able to name the helpers the same and have just those pieces in the individual runtime partitions.

lambdageek replied:
Getting some conflicting guidance here https://github.com/dotnet/runtime/pull/48380#discussion_r578846729 :smile: 

But actually maybe your suggestion was to just leave the QCall / InternallCall runtime-specific and to share the invocations.  Unfortunately that won't work today - the arguments are a bit different for each runtime.

stephentoub replied:
> maybe your suggestion was to just leave the QCall / InternallCall runtime-specific and to share the invocations

Yeah, that's what I was hoping for.  If it's not possible, ok.

marek-safar commented on src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:
```diff
@@ -1,11 +1,69 @@
+#endif
+        }
+
+        internal static void ApplyUpdateSdb(Assembly assembly, byte[] metadataDelta, byte[] ilDelta, byte[]? pdbDelta = null)
```
Why is this method needed?

lambdageek replied:
I don't know how to call a method taking a `ReadOnlySpan<byte>` argument using debugger-libs.

marek-safar replied:
Shouldn't this code be in debugger-libs when ApplyUpdate is public API?

lambdageek replied:
The issue is that we don't want to add a debugger-libs command for hot reload (we could do it, but previously there didn't seem to be a need).  So we're using the existing "create a byte[] blob" and "invoke a function" debugger-libs commands.  The "invoke a function" debugger-libs command can't use `ROS<byte>` arguments - at least I don't know how to do it.

So there are three options:
1. Call a function with better arguments (what i'm doing in this PR)
2. or, add a soft debugger protocol command for applying a hot reload delta.
3. Add a way to pass `ref struct` arguments when invoking using the debugger.

marek-safar replied:
Why cannot we do "create a byte[] blob" and "invoke a function" inside debugger-libs? Either way, this method cannot be in a permanent descriptor or the hot reload will always ship with release config

lambdageek replied:
We will probably need to add a new command to the debugger protocol, then.

Can I keep this for now?

marek-safar replied:
I'd rather not as it's gonna screw our size tracking numbers

lambdageek replied:
Discussed offline: this method is (temporarily) replacing an existing LoadMetadataUpdate private API so the size tracking isn't affected by this PR.  I added an issue to switch to debugger libs and remove ApplyUpdateSdb.  But we're going to keep it for now to unblock downstream work

marek-safar commented on src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:
```diff
@@ -1,11 +1,69 @@
+#if !FEATURE_METADATA_UPDATE
+            throw new NotSupportedException ("Method body replacement not supported in this runtime");
+#else
+            if (assembly == null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            RuntimeAssembly? runtimeAssembly = assembly as RuntimeAssembly;
+            if (runtimeAssembly == null)
+            {
+                throw new ArgumentException(SR.Argument_MustBeRuntimeAssembly);
+            }
```
```suggestion
		    if (assembly is not RuntimeAssembly runtimeAssembly)
		    {
			    if (assembly is null) throw new ArgumentNullException(nameof(assembly));
			    throw new ArgumentException(SR.Argument_MustBeRuntimeAssembly);
		    }
```

lambdageek committed:
Update src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs

Co-authored-by: Marek Safar <marek.safar@gmail.com>

lambdageek committed:
formatting

lambdageek committed:
more formatting

lambdageek committed:
Don't allow changes to System.Private.CoreLib

lambdageek committed:
Add a TODO to remove ApplyUpdateSdb

lambdageek committed:
update DeltaHelper

lambdageek committed:
don't use ApplyUpdateSdb in DeltaHelper

lambdageek commented:
@stephentoub  I rebased on master picking up @mikem8361 's changes and moved the argument checking to the shared area.  Would you mind reviewing again?

@marek-safar leaving ApplyUpdateSdb as we discussed for now.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:
```diff
@@ -0,0 +1,37 @@
+        /// <exception cref="NotSupportedException">The update could not be applied.</exception>
+        public static void ApplyUpdate(Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta)
+        {
+            if (assembly is not RuntimeAssembly runtimeAssembly)
```
I think it is contra-productive to be splitting a method into two to share 4 lines of trivial argument checking.

lambdageek replied:
Yep, will revert.  I misunderstood what Stephen was suggesting.

lambdageek committed:
Run the test on mono on all platoforms; CoreCLR windows only

mikem8361 commented:
@lambdageek, are you going to enable the invalid arg testing in runtime\src\libraries\System.Runtime.Loader\tests\AssemblyExtensionsTest.cs? By removing the "SkipOnMono" and "ActiveIssue" attributes?

FYI, to build/run the test on Windows:
```
call c:\runtime\dotnet.cmd build /t:Test /p:XunitMethodName=System.Reflection.Metadata.AssemblyExtensionsTest.ApplyUpdateInvalidParameters C:\runtime\src\libraries\System.Runtime.Loader\tests\System.Runtime.Loader.Tests.csproj
```
On Linux/MacOS:
```
$HOME/runtime/dotnet.sh build /t:Test /p:XunitMethodName=System.Reflection.Metadata.AssemblyExtensionsTest.ApplyUpdateInvalidParameters $HOME/runtime/src/libraries/System.Runtime.Loader/tests/System.Runtime.Loader.Tests.csproj
```

lambdageek commented:
Enabled the `ApplyUpdateInvalidParameters` test for Mono on all platforms.  ActiveIssue for CoreCLR on non-Windows. /cc @mikem8361

lambdageek commented:
It's unfortunate that `PlatformSpecificAttribute` doesn't have a runtime parameter.  ActiveIssue isn't really a great fit

mikem8361 commented on src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs:
```diff
@@ -12,9 +12,7 @@ class NonRuntimeAssembly : Assembly
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/45689", platforms: TestPlatforms.AnyUnix, runtimes: TestRuntimes.Mono)]
-        [PlatformSpecific(TestPlatforms.Windows)]
-        [SkipOnMono("Not yet implemented on Mono")]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/45689", platforms: ~TestPlatforms.Windows, runtimes: TestRuntimes.CoreCLR)]
```
I found out the hard way but ActiveIssue doesn't actually affect whether the test runs.  The test is going to fail on Linux/MacOS on coreclr

lambdageek replied:
I'm not totally convinced.  Just tried
```
        [Fact]
        [ActiveIssue("https://github.com/dotnet/runtime/issues/45689", platforms: ~TestPlatforms.Windows, runtimes: TestRuntimes.Mono)]
        public static void FailingTest()
        {
                Assert.True(false);
        }
```

and it skipped the test on Linux using Mono.  and when I changed to `TestPlatforms.CoreCLR` running on Mono on Linux ran the test and failed.

stephentoub replied:
> I found out the hard way but ActiveIssue doesn't actually affect whether the test runs. 

It should.  If it doesn't, that's a bug somewhere.

lambdageek replied:
runtime (Libraries Test Run release coreclr OSX x64 Debug) passed, so looks like it's working.

mikem8361 replied:
I must have had things backwards or something.

lambdageek committed:
throw InvalidOpertionException for corlib changes with hot reload disabled

stephentoub commented on src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.Mono.cs:
```diff
@@ -0,0 +1,47 @@
+#else
+            // System.Private.CoreLib is not editable
+            if (runtimeAssembly == typeof(AssemblyExtensions).Assembly)
+                throw new InvalidOperationException ("The assembly can not be edited or changed.");
```
Should these strings be in the resx?

lambdageek replied:
fixed

lambdageek committed:
Use resource string

lambdageek commented:
Is there anything else that should be done before this is merged?  If not could someone approve.

stephentoub commented on src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:
```diff
@@ -1,13 +1,68 @@
+                throw new InvalidOperationException (SR.InvalidOperation_AssemblyNotEditable);
+
+#if !FEATURE_METADATA_UPDATE
+            throw new NotSupportedException ("Method body replacement not supported in this runtime");
```
Not sure if this message was left in the code on purpose or should be extracted to the .resx.

stephentoub commented on src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:
```diff
@@ -1,13 +1,68 @@
+        /// <param name="pdbDelta">The PDB changes to be applied.</param>
+        /// <exception cref="ArgumentNullException">The assembly argument is null.</exception>
+        /// <exception cref="NotSupportedException">The update could not be applied.</exception>
+        public static void ApplyUpdate(Assembly assembly, ReadOnlySpan<byte> metadataDelta, ReadOnlySpan<byte> ilDelta, ReadOnlySpan<byte> pdbDelta = default)
```
The `= default` should be removed I believe.

stephentoub commented on src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:
```diff
@@ -1,13 +1,68 @@
+#endif
+        }
+
+        internal static void ApplyUpdateSdb(Assembly assembly, byte[] metadataDelta, byte[] ilDelta, byte[]? pdbDelta = null)
```
Is the `= null` needed here?

lambdageek committed:
Address review feedback
